cro-clib/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+pub mod bindings;
+use crate::bindings::{ApiContext, ApiResult, Buf, HDWallet, MAX_STRING_LENGTH};
+use chain_core::init::network::init_chain_id;
+use client_common::error::ErrorKind;
+use client_common::SecKey;
+use client_core::types::WalletKind;
+use client_core::wallet::{DefaultWalletClient, WalletClient};
+use client_core::HDSeed;
+use client_core::Mnemonic;
+use secstr::SecUtf8;
+use std::ffi::CStr;
+use std::str::FromStr;
+
+use client_common::storage::MemoryStorage;
+#[derive(Default)]
+pub struct ApiInfo {
+    pub chain_id: String,
+}
+static mut G_DB: Option<MemoryStorage> = None;
+
+/// # Safety
+/// this function should not be called inside rust
+pub unsafe fn get_string(src: *const ::std::os::raw::c_char) -> String {
+    CStr::from_ptr(src).to_string_lossy().into_owned()
+}
+
+pub fn copy_string(src: &str, dst: &mut [u8]) {
+    dst[..src.len()].clone_from_slice(&src.as_bytes()[..src.len()]);
+    dst[src.len()] = 0;
+}
+
+pub fn get_string_from_buf(src: &Buf) -> String {
+    std::str::from_utf8(&src.buf[0..src.length as usize])
+        .expect("convert to string")
+        .to_string()
+}
+
+fn get_buf(src: &str) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    copy_string(src, &mut ret.buf);
+    ret.length = src.len() as i32;
+    ret
+}
+
+fn get_buf_vec(src: &[u8]) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    ret.buf[..src.len()].clone_from_slice(&src[..]);
+    ret.length = src.len() as i32;
+    ret
+}
+
+#[no_mangle]
+pub fn get_network_id() -> ApiResult {
+    let networkid = chain_core::init::network::get_network_id();
+    let networkid_string = format!("{:2X}", networkid);
+    ApiResult {
+        error: 0,
+        value: get_buf(&networkid_string),
+    }
+}
+
+#[no_mangle]
+pub fn create_wallet() -> ApiResult {
+    ApiResult {
+        error: 0,
+        value: get_buf("wallet"),
+    }
+}
+
+/// # Safety
+/// this function should not be called inside rust
+pub unsafe fn do_create_hdwallet(
+    name: *const ::std::os::raw::c_char,
+    passphrase: *const ::std::os::raw::c_char,
```

Comment:
`use std::os::raw::c_char`?

Reply:
c-bindgen doesn't process struct,
output is the same.
i'll change to use heap.